src/memtest.rs

```diff
@@ -0,0 +1,380 @@
+            timeout_checker.check()?;
+            let val = if j % 2 == 0 { val } else { !val };
+            write_volatile(base_ptr.add(j), val);
+            write_volatile(half_ptr.add(j), val);
+        }
+        compare_regions(base_ptr, half_ptr, half_count, &mut timeout_checker)?;
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+pub unsafe fn test_block_seq(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_ms: usize,
+) -> Result<MemtestOutcome, MemtestError> {
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let mut timeout_checker = TimeoutChecker::new(Instant::now(), timeout_ms, half_count * 2 * 256);
+
+    for i in 0..=255 {
+        let mut val: usize = 0;
+        write_bytes(&mut val, i, 1);
+        for j in 0..half_count {
+            timeout_checker.check()?;
+            write_volatile(base_ptr.add(j), val);
+            write_volatile(half_ptr.add(j), val);
+        }
+        compare_regions(base_ptr, half_ptr, half_count, &mut timeout_checker)?;
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+impl TimeoutChecker {
+    fn new(start_time: Instant, timeout_ms: usize, total_iter: usize) -> TimeoutChecker {
+        TimeoutChecker {
+            start_time,
+            timeout_ms,
+            total_iter,
+            completed_iter: 0,
+            checkpoint: 1,
+            num_checks_completed: 0,
+            // TODO: Choice of starting interval is arbitrary for now.
+            checking_interval_ns: 1000.0,
+        }
+    }
+
+    /// This function should be called in every iteration of a memtest
+    ///
+    /// To minimize overhead, the time is only checked at specific checkpoints.
+    /// The algorithm makes a prediction of how many iterations will be done during `checking_interval_ns`
+    /// and sets the checkpoint based on the prediction.
+    ///
+    /// The algorithm also determines whether it is likely that the test will be completed
+    /// based on `work_progress` and `time_progress`.
+    /// If it is likely that the test will be completed, `checking_interval_ns` is scaled up to be more
+    /// lenient and reduce overhead.
+    fn check(&mut self) -> Result<(), MemtestError> {
+        if self.completed_iter < self.checkpoint {
+            self.completed_iter += 1;
+            return Ok(());
+        }
+
+        let elapsed = self.start_time.elapsed();
+        if elapsed > Duration::from_millis(self.timeout_ms as u64) {
+            return Err(MemtestError::Timeout);
+        }
+
+        let work_progress = self.completed_iter as f64 / self.total_iter as f64;
+        let time_progress = elapsed.as_millis() as f64 / self.timeout_ms as f64;
+        // TODO: Consider having a max for `checking_interval_ns` to have a reasonable timeout guarantee
+        if work_progress > time_progress {
+            self.checking_interval_ns *= 2.0;
+        }
+
+        let avg_iter_time_ns = elapsed.as_nanos() as f64 / self.completed_iter as f64;
```

Comment:
Also, since you are still early in your Rust career, do yourself a favour and forget that you ever knew you could cast integers with the `as` keyword. The Rust Core team [wants to deprecate it](https://internals.rust-lang.org/t/lets-deprecate-as-for-lossy-numeric-casts/16283).

The TL;DR for why: The other operations like add/sub/mul/div all have versions that explicitly specify how to handle overflow (Ex. `checked_add()`, `wrapping_add()` and `saturating_add()`). And they all statically guarantee that you are not accidentally casting to a type more narrow than the source. This is important for Rust, a language that tries to encourage you to do things that are safe and requires you to be explicit when you're doing something that might accidentally cause a bug.

`as` casts have none of that -- they always wrap, and they will even let you narrow and wrap at the same time. For example, let's say that you have code where you `as` casted from a `u64` to a `usize`. Maybe it works perfectly on 64-bit OS, but suddenly you try it on 32-bit and your code stops working -- Oops! You overflowed the `usize` (which is now only 32-bits) and the compiler wrapped the value without any warning.

Instead of `as` casting, use `From` and `TryFrom` traits. `From` will always be available if the cast can't fail, and `TryFrom` for the casts that could.

Examples:

```Rust
let age = 82u8;
let big_age = u32::from(age); // Will work fine -- A u8 can always be represented as a u32

let population = 4_000_000_000u64;
// let small_population = u32::from(population); // Won't work, u32 can't fit all u64 values
let small_population = u32::try_from(population).unwrap(); // Works fine. Will panic if population rises to 2^32 though.
```